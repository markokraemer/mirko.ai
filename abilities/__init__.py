from abilities.abilities import TerminalOps
from abilities.abilities import RetrievalOps
from abilities.abilities import OpsResult
from abilities.abilities import get_container_merged_dir
