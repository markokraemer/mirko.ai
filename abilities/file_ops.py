import os
import json
import logger

excluded_files = ['.DS_Store', '.gitignore', 'package-lock.json', 'postcss.config.js', 'tailwind.config.js', 'next.config.js', 'playwright.config.js', 'jsconfig.json', 'components.json', 'API_doc.md', 'env.example']
excluded_dirs = ['cypress', 'cypress/downloads', 'node_modules', 'migrations', '.next', 'cypress/screenshots', 'playwright-report', 'test-results', 'dist', 'build', 'coverage' ] 
excluded_file_extensions = ['.ico', '.svg', '.png', '.jpg','.jpeg', '.gif', '.bmp', '.tiff', '.webp']


def read_directory(directory):
    """Reads all files in a given directory and returns a file tree and file contents, excluding certain files and directories."""
    # logger.info(f"Reading directory {directory}")
    file_tree = []
    file_contents = {}

    for root, dirs, files in os.walk(directory):
        dirs[:] = [d for d in dirs if d not in excluded_dirs]
        files = [f for f in files if f not in excluded_files and not any(
            f.endswith(ext) for ext in excluded_file_extensions)]

        root_path = os.path.relpath(root, directory)
        file_tree.append({"directory": root_path, "files": files})

        for file in files:
            file_path = os.path.join(root, file)
            try:
                with open(file_path, 'r', encoding='utf-8') as file_content:
                    file_contents[file_path] = file_content.read()
            except UnicodeDecodeError:
                # logger.info(f"Skipping non-UTF-8 encoded file: {file_path}")
                continue

    return file_tree, file_contents

def read_file_content(file_path):
    if os.path.exists(file_path):
        with open(file_path, 'r') as file:
            return file.read()
    # logger.info(f"Reading file content from {file_path}")
    return ""
    
def modify_file(file_path, new_content):
    # Ensure the directory exists
    os.makedirs(os.path.dirname(file_path), exist_ok=True)

    # Create or overwrite the file with the new content
    with open(file_path, 'w') as file:
        file.write(new_content)
    logger.info(f"Writing new content to {file_path}")

def rename_file(file_path, new_name):
    new_file_path = os.path.join(os.path.dirname(file_path), new_name)
    os.rename(file_path, new_file_path)
    logger.info(f"Renamed file {file_path} to {new_file_path}")

def delete_file(file_path):
    os.remove(file_path)
    logger.info(f"Deleted file {file_path}")

def get_file_paths_by_name(file_name, directory):
    # Search for the file in the given directory and collect all matches
    matching_file_paths = []
    for root, dirs, files in os.walk(directory):
        if file_name in files:
            matching_file_paths.append(os.path.join(root, file_name))
    if not matching_file_paths:
        raise FileNotFoundError(f"File {file_name} not found in {directory}.")
    return matching_file_paths
