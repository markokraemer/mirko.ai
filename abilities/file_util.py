import os
from pathlib import Path

EXCLUDED_FILES = ['.DS_Store', '.gitignore', 'package-lock.json', 'postcss.config.js', 'tailwind.config.js', 'next.config.js', 'playwright.config.js', 'jsconfig.json', 'components.json', 'API_doc.md', 'env.example']
EXCLUDED_DIRS = ['ui', 'cypress', 'node_modules', 'migrations', '.next', 'playwright-report', 'test-results', 'dist', 'build', 'coverage' ] 
EXCLUDED_EXT = ['.ico', '.svg', '.png', '.jpg','.jpeg', '.gif', '.bmp', '.tiff', '.webp']


def _should_exclude(root_path, path):
    """
    Check if a file or directory should be excluded based on the exclusions and depth.
    """
    for dir_name in EXCLUDED_DIRS:
        if os.path.commonpath([path, os.path.join(root_path, dir_name)]) == os.path.normpath(os.path.join(root_path, dir_name)):
            return True
    return False

def find_files(root_path, depth):
    """
    Recursively find files in the given directory up to a certain depth, 
    excluding specified files, directories, and file extensions.
    """
    found_files = []
    
    for root, dirs, files in Path(root_path).walk():
        # Calculate current depth
        current_depth = root.relative_to(root_path).parts
        if len(current_depth) > depth:
            break
        for file in files:
            if file in EXCLUDED_FILES or any(file.endswith(ext) for ext in EXCLUDED_EXT):
                continue
            file_path = root / file
            if not _should_exclude(root_path, file_path):
                found_files.append(str(file_path.relative_to(root_path)))



    return found_files
