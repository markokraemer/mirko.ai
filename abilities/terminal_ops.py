# __init__.py

import docker
from dataclasses import dataclass


@dataclass
class OpsResult:
    ok: bool
    exit_code: int
    output: bytes


class TerminalOps:
    def __init__(self, container_name: str):
        self.client = docker.from_env()
        self.container = self.client.containers.get(container_name)

    def send_command(self, command: str) -> OpsResult:
        """
        Run command in the terminal and wait for the result
        """
        result = self.container.exec_run(command)
        return OpsResult(
            ok=(result.exit_code == 0), output=result.output, exit_code=result.exit_code
        )

    @staticmethod
    def schema() -> list[dict]:
        """
        Returns the JSON schema for OpenAI function calls.
        """
        return [
            {
                "name": TerminalOps.send_command.__name__,
                "description": TerminalOps.send_command.__doc__,
                "parameters": {
                    "type": "object",
                    "properties": {
                        "command": {
                            "type": "string",
                            "description": "The command to be executed in the terminal.",
                        }
                    },
                    "required": ["command"],
                },
            }
        ]
