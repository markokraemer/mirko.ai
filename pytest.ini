[pytest]
testpaths = tests
addopts = "--approvaltests-use-reporter='PythonNativeReporter'"
