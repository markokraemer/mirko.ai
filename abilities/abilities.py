from dataclasses import dataclass
from abc import ABC, abstractmethod
import os
import json
import docker
from pathlib import Path
from enum import Enum
from xml.dom.minidom import getDOMImplementation

from abilities import file_util

__XML_DOM__ = getDOMImplementation()

@dataclass
class OpsResult:
    ok: bool
    exit_code: int
    output: str


class Ops(ABC):
    @classmethod
    @abstractmethod
    def openai_schema() -> list[dict]:
        """
        Returns the JSON schema for OpenAI function calls.
        """
        pass

    @classmethod
    def anthropic_schema(cls) -> str:
        """
        Return Anthropic Claude tool definition XML as a string
        https://docs.anthropic.com/claude/docs/functions-external-tools
        """
        doc = __XML_DOM__.createDocument(None, "tools", None)
        for tool_dict in cls.openai_schema():
            desc_node = doc.createElement("tool_description")
            tool_name_node = doc.createElement("tool_name")
            tool_name_node.appendChild(doc.createTextNode(tool_dict['function']['name']))
            desc_node.appendChild(tool_name_node)

            description_node = doc.createElement("description")
            description_node.appendChild(doc.createTextNode(tool_dict['function']['description']))
            desc_node.appendChild(description_node)

            parameters_node = doc.createElement("parameters")
            for param_name, param in tool_dict['function']['parameters']['properties'].items():
                parameter_node = doc.createElement("parameter")

                name_node = doc.createElement("name")
                name_node.appendChild(doc.createTextNode(param_name))
                parameter_node.appendChild(name_node)

                type_node = doc.createElement("type")
                type_node.appendChild(doc.createTextNode(param['type']))
                parameter_node.appendChild(type_node)

                description_node = doc.createElement("description")
                description_node.appendChild(doc.createTextNode(param['description']))
                parameter_node.appendChild(description_node)

                parameters_node.appendChild(parameter_node)

            desc_node.appendChild(parameters_node)
            doc.documentElement.appendChild(desc_node)
        return doc.toxml()

    def ok_response(self, data: dict | str) -> OpsResult:
        """
        Create an successfull OpsResult with data as the output json.
        """
        if isinstance(data, str):
            text = data
        else:
            text = json.dumps(data, indent=2)
        return OpsResult(ok=True, exit_code=0, output=text)

    def fail_response(self, msg: str, code: int = 1) -> OpsResult:
        """
        Create an successfull OpsResult with data as the output json.
        """
        return OpsResult(ok=False, exit_code=code, output=msg)


class TerminalOps(Ops):
    def __init__(self, container_name: str):
        self.client = docker.from_env()
        self.container = self.client.containers.get(container_name)

    def send_command(self, command: str) -> OpsResult:
        """
        Run command in the terminal and wait for the result
        """
        result = self.container.exec_run(command)
        # Since output is decoded to string, commands with binary output will not work.
        return OpsResult(
            ok=(result.exit_code == 0),
            output=str(result.output),
            exit_code=result.exit_code,
        )

    @classmethod
    def openai_schema(cls) -> list[dict]:
        """
        Returns the JSON schema for OpenAI function calls.
        """
        return [
            {
                "type": "function",
                "function": {
                    "name": TerminalOps.send_command.__name__,
                    "description": TerminalOps.send_command.__doc__,
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "command": {
                                "type": "string",
                                "description": "The command to be executed in the terminal.",
                            }
                        },
                        "required": ["command"],
                    },
                },
            }
        ]


def get_container_merged_dir(container_name: str):
    """
    Lookup workspace dir of docker container.
    Has a fallback for OrbStack, there may be other possibilities.
    It's not clear if there's a reliable way to get this.
    """
    fallback = os.path.expanduser(f"~/OrbStack/docker/containers/{container_name}")
    client = docker.from_env()
    container = client.containers.get(container_name)
    container_info = container.attrs
    merged_dir = container_info["GraphDriver"]["Data"]["MergedDir"]
    if os.path.exists(merged_dir):
        return merged_dir
    if os.path.exists(fallback):
        return fallback
    raise Exception(f"Cannot find {container_name} container workspace dir")


def _rindex(li, value):
    return len(li) - li[-1::-1].index(value) - 1


class RetrievalOps(Ops):
    def __init__(self, base_path: str):
        self.base_path = base_path
        # The docstrings are what the agent sees.
        # When they say relative to the current dir, that means be the current dir
        # in the container terminal session once we are wired up.

    def _get_effective_path(self, path: str) -> str:
        # Handling . and ~ to avoid touching things outside base_path.
        path_parts = path.split(os.sep)
        if "." in path_parts:
            remaining_path = path_parts[_rindex(path_parts, ".") + 1 :]
            effective_path = os.path.join(self.base_path, *remaining_path)
        else:
            effective_path = os.path.join(self.base_path, path)
        return effective_path

    def get_file_tree(self, path: str, depth: int = 1) -> OpsResult:
        """
        List path recursively with limited depth. Path is relative to current dir.
        """
        if not os.path.exists(self.base_path):
            return self.fail_response("Base path does not exist")

        if "~" in path:
            return self.fail_response("Path cannot contain ~")
        effective_path = self._get_effective_path(path)
        # This will catch any remaining case that could get outside base_path
        if not os.path.commonpath([self.base_path, effective_path]) == os.path.normpath(
            self.base_path
        ):
            return OpsResult(
                ok=False, output="Path is not within the base path", exit_code=1
            )
        try:
            listed_paths = file_util.find_files(effective_path, depth)
            return self.ok_response({"paths": listed_paths})
        except Exception as e:
            return OpsResult(ok=False, output=str(e), exit_code=1)

    def read_file_contents(self, path: str) -> OpsResult:
        """
        Show the contents of the file at path relative to current dir
        """
        if not os.path.exists(self.base_path):
            return self.fail_response("Base path does not exist")
        effective_path = self._get_effective_path(path)
        print(effective_path)
        # Make sure we didn't break out of base_path
        if not os.path.commonpath([self.base_path, effective_path]) == os.path.normpath(
            self.base_path
        ):
            return OpsResult(
                ok=False, output="Path is not within the base path", exit_code=1
            )
        if os.path.exists(effective_path):
            with open(effective_path, "r") as file:
                text = file.read()
            return self.ok_response(text)
        else:
            return self.fail_response("File does not exist")

    @classmethod
    def openai_schema(cls) -> list[dict]:
        """
        Returns the JSON schema for OpenAI function calls.
        """
        return [
            {
                "type": "function",
                "function": {
                    "name": RetrievalOps.get_file_tree.__name__,
                    "description": RetrievalOps.get_file_tree.__doc__,
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "path": {
                                "type": "string",
                                "description": "Directory to list relative to current",
                            },
                            "depth": {
                                "type": "number",
                                "description": "Depth limit, default 1",
                            },
                        },
                        "required": ["path"],
                    },
                },
            },
            {
                "type": "function",
                "function": {
                    "name": RetrievalOps.read_file_contents.__name__,
                    "description": RetrievalOps.read_file_contents.__doc__,
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "path": {
                                "type": "string",
                                "description": "Path of file relative to current dir",
                            }
                        },
                        "required": ["path"],
                    },
                },
            },
        ]
