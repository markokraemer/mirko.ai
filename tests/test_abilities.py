import os
import pytest
from abilities import TerminalOps, RetrievalOps, get_container_merged_dir
from abilities.file_util import find_files
from approvaltests.approvals import verify

def assert_tool_fields(schema):
    """
    Check some basic fields on an OpenAI function call tool definition.
    If we wanted to be rigorious we could use the jsonschema, but this is a start.
    """
    for tool in schema: 
        assert tool['type'] == 'function'
        assert tool['function']
        assert tool['function']['name']
        assert tool['function']['description']
        assert tool['function']['parameters']

def test_terminal_schema():
    subject = TerminalOps.schema()
    assert_tool_fields(subject)
    verify(subject)

def test_retrieval_schema():
    subject = RetrievalOps.schema()
    assert_tool_fields(subject)
    verify(subject)


def test_find_files(fs):
    # The fs fixture is pyfakefs
    fs.create_file("/var/fakedata/a/xx1.txt")
    fs.create_file("/var/fakedata/a/.DS_Store") # Excluded
    
    found = find_files("/var/fakedata", 1)
    assert found == ["a/xx1.txt"]

def test_find_files_depth_limit(fs):
    fs.create_file("/var/fakedata/a/b/xx1.txt")
    found = find_files("/var/fakedata", 1)
    assert found == []

def test_find_files_exclude_ext(fs):
    fs.create_file("/var/fakedata/xx1.jpg")
    found = find_files("/var/fakedata", 1)
    assert found == []

def test_find_files_exclude_dirname(fs):
    fs.create_file("/var/fakedata/node_modules/xx1.jpg")
    found = find_files("/var/fakedata", 3)
    assert found == []



if os.getenv("INTEGRATION"):
    def test_send_ls():
        print(TerminalOps("workspace-dev-env-1").send_command("ls"))

    def test_list_tree():
        base_path = get_container_merged_dir("workspace-dev-env-1")
        print(base_path)
        subject = RetrievalOps(base_path)
        print(subject.get_file_tree(".", 2))
        