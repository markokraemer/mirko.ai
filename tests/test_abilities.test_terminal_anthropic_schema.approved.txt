<?xml version="1.0" ?><tools><tool_description><tool_name>send_command</tool_name><description>
        Run command in the terminal and wait for the result
        </description><parameters><parameter><name>command</name><type>string</type><description>The command to be executed in the terminal.</description></parameter></parameters></tool_description></tools>
